pkg/instancegroups/duplicate.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instancegroups
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kops/pkg/apis/kops"
+)
+
+// duplicateProvider session data
+type duplicateProvider struct {
+	GroupUpdate *RollingUpdateInstanceGroup
+}
+
+// NewDuplicateRollout creates and returns a drain provider
+func NewDuplicateRollout(update *RollingUpdateInstanceGroup) Rollout {
+	return &duplicateProvider{GroupUpdate: update}
+}
+
+// RollingUpdate performs a duplicate rolling update of an instance group
+func (p *duplicateProvider) RollingUpdate(ctx context.Context, list *api.InstanceGroupList) error {
+	group := p.GroupUpdate.CloudGroup
+	ig := group.InstanceGroup
+	name := group.InstanceGroup.Name
+	strategy := ig.Spec.Strategy
+	update := p.GroupUpdate.Update
+
+	update.Infof("rolling out with strategy: %s, instancegroups: %s, batch: %d", strategy.Name, name, strategy.Batch)
+
+	// @check the instancegroup is not already a duplicate
+	if _, found := ig.GetAnnotations()[DuplicateParentAnnotation]; found {
+		update.Infof("ignoring instancegroup: %s as we found parent duplicate annotation", ig.Name)
+		return nil
+	}
+
+	// @step: attempt to duplicate the instance group
+	newName := fmt.Sprintf("%s-%d", group.InstanceGroup.Name, time.Now().Unix())
```

Comment:
So we had a similar PR from chrislovecnm, and we were thinking of instead introducing a new object... if Nodes are Pods, InstanceGroups would be Deployments, and we're thinking of creating the ReplicaSet.  https://github.com/kubernetes/kops/pull/2818#issuecomment-320138678

This seems even more likely now with the work happening happening on Machines & MachineSets - https://github.com/kubernetes/kube-deploy/tree/master/cluster-api

So an InstanceGroup update would do what I think you're doing here, except that we would create a MachineSet

There's also an implementation plan we've been discussing where we put this logic into a machine controller.

We should probably discuss :-)